pkg/yttlibrary/regexp.go

```diff
@@ -47,3 +48,67 @@ func (b regexpModule) Match(thread *starlark.Thread, f *starlark.Builtin, args s
 
 	return starlark.Bool(matched), nil
 }
+
+func (b regexpModule) Replace(thread *starlark.Thread, f *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	if args.Len() != 3 {
+		return starlark.None, fmt.Errorf("expected exactly 3 arguments")
+	}
+
+	pattern, err := core.NewStarlarkValue(args.Index(0)).AsString()
+	if err != nil {
+		return starlark.None, err
+	}
+
+	re, err := regexp.Compile(pattern)
+	if err != nil {
+		return starlark.None, err
+	}
+
+	source, err := core.NewStarlarkValue(args.Index(1)).AsString()
+	if err != nil {
+		return starlark.None, err
+	}
+
+	repl := args.Index(2)
+	switch repl.(type) {
+	case *starlark.Function:
+		return b.replaceLambda(thread, re, source, repl.(*starlark.Function))
+	default:
+		return b.replaceString(re, source, repl)
+	}
+}
+
+func (b regexpModule) replaceString(re *regexp.Regexp, source string, repl starlark.Value) (starlark.Value, error) {
+	replStr, err := core.NewStarlarkValue(repl).AsString()
+	if err != nil {
+		return starlark.None, err
+	}
+
+	newString := re.ReplaceAllString(source, replStr)
+
+	return starlark.String(newString), nil
+}
+
+func (b regexpModule) replaceLambda(thread *starlark.Thread, re *regexp.Regexp, source string, repl *starlark.Function) (result starlark.Value, err error) {
+	defer func() {
+		if r, ok := recover().(error); ok {
+			err = r
```

Comment:
"Catching" the error here to avoid an ugly stack trace if there's an error evaluating the applied lambda function or casting the result to a string.

Example:

```yaml
#@ load("@ytt:regexp", "regexp")
test: #@ regexp.replace("[a-z]+[0-9]+", "__hello123__", lambda s: len(s))
```

```
ytt: Error:
- regexp.replace: expected starlark.String, but was starlark.Int
    in <toplevel>
      test.yml:2 | test: #@ regexp.replace("[a-z]+[0-9]+", "__hello123__", lambda s: len(s))
```

Reply:
hmm, i wonder if there is a better way to propagate errors that happen instead of using panic/recover... what if we have `var lastErr error` and we set it to error that happens (if value isnt string), and then for all subsequent calls to replace function if err is non-nil return empty string. then at the end return lastErr.

Reply:
That approach works, although my implementation (using named return params in the enclosing function) doesn't quite adopt the separate `lastErr` variable. Lmk what you think.